modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersAuthenticator.java

```diff
@@ -0,0 +1,149 @@
+import org.elasticsearch.action.support.ContextPreservingActionListener;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.http.BasicHttpRequest;
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+public class HttpHeadersAuthenticator {
+
+    private final TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext;
+    private final BiConsumer<BasicHttpRequest, ActionListener<Void>> authenticate;
+    private final ThreadContext threadContext;
+
+    public HttpHeadersAuthenticator(
+        TriConsumer<BasicHttpRequest, Channel, ThreadContext> populateThreadContext,
+        BiConsumer<BasicHttpRequest, ActionListener<Void>> authenticate,
+        ThreadContext threadContext
+    ) {
+        this.populateThreadContext = populateThreadContext;
+        this.authenticate = authenticate;
+        this.threadContext = threadContext;
+    }
+
+    public static final HttpHeadersAuthenticator NOOP = new HttpHeadersAuthenticator(null, null, null) {
+        @Override
+        public DefaultHttpRequest wrapNewMessage(DefaultHttpRequest decodedNewMessage) {
+            return decodedNewMessage;
+        }
+
+        @Override
+        public void authenticateMessage(HttpRequest request, Channel channel, ActionListener<Void> listener) {
+            listener.onResponse(null);
+        }
+    };
+
+    public DefaultHttpRequest wrapNewMessage(DefaultHttpRequest decodedNewMessage) {
+        HttpHeadersWithAuthenticationContext httpHeadersWithAuthenticationContext = new HttpHeadersWithAuthenticationContext(
+            decodedNewMessage.headers()
+        );
+        return new DefaultHttpRequest(
+            decodedNewMessage.protocolVersion(),
+            decodedNewMessage.method(),
+            decodedNewMessage.uri(),
+            httpHeadersWithAuthenticationContext
+        );
+    }
+
+    public void authenticateMessage(HttpRequest request, Channel channel, ActionListener<Void> listener) {
+        assert request.headers() instanceof HttpHeadersAuthenticator.HttpHeadersWithAuthenticationContext;
+        final BasicHttpRequest requestToAuthenticate = wrapToBasicHttpRequest(request);
+        final Supplier<ThreadContext.StoredContext> emptyContext = threadContext.wrapRestorable(threadContext.newStoredContext());
+        final ActionListener<Void> contextPreservingListener = new ContextPreservingActionListener<>(emptyContext, listener);
+        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            populateThreadContext.apply(requestToAuthenticate, channel, threadContext);
+            authenticate.accept(requestToAuthenticate, ActionListener.wrap(ignored -> {
+                final ThreadContext.StoredContext authenticatedContext = threadContext.newStoredContext();
```

Comment:
Following a successful authentication the thread context gets into an authenticated state.
This captures the full authenticated thread context, to be reinstanted downstream, when executing the  http handler.

Reply:
I think I remember you said that we need to associate this authenticated context with the request since the context will get lost since it may not part of the same call stack... but can you detail abit more why that is ? Isn't both this code (storing the authenticated context) and the code that restores the authenticated context happening on the same eventloop thread ?  (also, what thread does the current authentication happen on ->  the eventloop thread?)

Reply:
> Isn't both this code (storing the authenticated context) and the code that restores the authenticated context happening on the same eventloop thread ?

This is a great point to raise!
Yes, that is correct, both pieces of code (setting in and then upacking) run under the same thread.
The fundamental issue is with the call stack. I'll explain briefly but this might be worth some sync conv.
The authenticated context needs to be set for a specific request only. This is crucial. The way we normally achieve this is to set it in a try-with-resource block and, just after that, inside the block, we call the handler for the request. When the handler for that request returns, the try-block closes, and the thread-context is restored to the un-authenticated state. Hence, there is no danger that a single handler being passed-in a single request, somehow starts executing another request.
Contrast this to setting the authenticated context earlier in the process, in the netty channel handler. The concept of a full request is not there yet, there are only request pieces. A good guess is to set an authenticated thread-context when the last piece of a HTTP request is forwarded. This relies on the fact that the same call stack that forwards the last piece is going to invoke the handler for the completely assembled request (I also tested it and works). The danger, of course, is that theoretically the full-request might be buffered somewhere in a downstream netty channel handler or ES code. In this case there might be multiple requests that are executed under a single authenticated thread-context.

Reply:
> what thread does the existing (before this PR) authentication happen on -> the eventloop thread?

Yes the same eventloop thread does the authentication currently too.

Reply:
> I'll explain briefly but this might be worth some sync conv.

Thanks, I scheduled a discussion on Monday.